File: EDUCATIVE/TOP_K_HEAP/FrequencySort.java
import java.util.*;

class FrequencySort {

    public static String sortCharacterByFrequency(String str) {
        // TODO: Write your code here

        char [] chars = str.toCharArray();
        HashMap<Character, Integer> freqSort = new HashMap<>();

        for(char ch : chars) {
            freqSort.put(ch, freqSort.getOrDefault(ch, 0) + 1);
        }

        PriorityQueue<Map.Entry<Character, Integer>> maxHeap = new PriorityQueue<Map.Entry<Character, Integer>>((o1, o2) -> o2.getValue() - o1.getValue());

        for(Map.Entry<Character, Integer> entry: freqSort.entrySet()) {
            maxHeap.add(entry);
        }

        StringBuilder result = new StringBuilder();

        while(!maxHeap.isEmpty()) {
            int count = maxHeap.peek().getValue();
            while(count > 0) {
                result.append(maxHeap.peek().getKey());
                count--;
            }
            maxHeap.poll();
        }

       
        return result.toString();
    }

    public static void main(String[] args) {
        String result = FrequencySort.sortCharacterByFrequency("Programming");
        System.out.println("Here is the given string after sorting characters by frequency: " + result);

        result = FrequencySort.sortCharacterByFrequency("abcbab");
        System.out.println("Here is the given string after sorting characters by frequency: " + result);
    }
}

File: EDUCATIVE/TOP_K_HEAP/frequency_sort.py
from collections import defaultdict
from heapq import *

def sort_character_by_frequency(str):
    freq_str = defaultdict(int)

    for i in str:
        freq_str[i] += 1
    
    maxHeap = []
    for char, freq in freq_str.items():
        heappush(maxHeap, (-freq, char))
    
    sorted_string = []
    
    while maxHeap:
        freq, char = heappop(maxHeap)
        #print ("{0} and {1}".format(freq, char))
        for _ in range(-freq):
            sorted_string.append(char)

    return ''.join(sorted_string)


def main():
    print("String after sorting characters by frequency: " +
        sort_character_by_frequency("Programming"))
    print("String after sorting characters by frequency: " +
        sort_character_by_frequency("abcbab"))

main()